numpy/_core/src/multiarray/stringdtype/casts.c

```diff
@@ -1503,13 +1507,258 @@ STRING_TO_CFLOAT_CAST(clongdouble, l, longdouble)
+        if (given_descrs[1]->names != NULL || given_descrs[1]->subarray != NULL) {
+            PyErr_SetString(
+                    PyExc_TypeError,
+                    "Casting from StringDType to a structured dtype is not "
+                    "supported.");
+            return (NPY_CASTING)-1;
+        }
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    return NPY_UNSAFE_CASTING;
+}
+
+static int
+string_to_void(PyArrayMethod_Context *context, char *const data[],
+               npy_intp const dimensions[], npy_intp const strides[],
+               NpyAuxData *NPY_UNUSED(auxdata))
+{
+    StringDTypeObject *descr = (StringDTypeObject *)context->descriptors[0];
+    npy_string_allocator *allocator = NpyString_acquire_allocator(descr);
+    int has_null = descr->na_object != NULL;
+    int has_string_na = descr->has_string_na;
+    const npy_static_string *default_string = &descr->default_string;
+    const npy_static_string *na_name = &descr->na_name;
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+    size_t max_out_size = context->descriptors[1]->elsize;
+
+    while (N--) {
+        const npy_packed_static_string *ps = (npy_packed_static_string *)in;
+        npy_static_string s = {0, NULL};
+        int is_null = NpyString_load(allocator, ps, &s);
+        if (is_null == -1) {
+            npy_gil_error(PyExc_MemoryError,
+                          "Failed to load string in void to string cast");
+            goto fail;
+        }
+        else if (is_null) {
+            if (has_null && !has_string_na) {
+                // lossy but not much else we can do
+                s = *na_name;
+            }
+            else {
+                s = *default_string;
+            }
+        }
+
+        // This might truncate a UTF-8 character. Should we warn if that
+        // happens?  UTF-8 won't be round-trippable if there is truncation
+        memcpy(out, s.buf, s.size > max_out_size ? max_out_size : s.size);
+        if (s.size < max_out_size) {
+            memset(out + s.size, 0, (max_out_size - s.size));
+        }
+
+        in += in_stride;
+        out += out_stride;
+    }
+
+    NpyString_release_allocator(descr);
+
+    return 0;
+
+fail:
+
+    NpyString_release_allocator(descr);
+
+    return -1;
+}
+
+static PyType_Slot s2v_slots[] = {
+        {NPY_METH_resolve_descriptors, &string_to_void_resolve_descriptors},
+        {NPY_METH_strided_loop, &string_to_void},
+        {0, NULL}};
+
+static char *s2v_name = "cast_StringDType_to_void";
+
+// void to string
+
+// Bjoern Hoerhmann DFA UTF-8 validator
+// original code has an MIT license
+// Copyright (c) 2008-2009 Bjoern Hoehrmann <bjoern@hoehrmann.de>
+// See http://bjoern.hoehrmann.de/utf-8/decoder/dfa/ for details.
```

Comment:
I guess this is enough to be only in the code?

Reply:
Added the license text too to be safe. I see we do similar things with other snippets extracted from compatibly licensed open source code, no need to update the bundled licenses file.